### ISSUE TITLE
Sort non-numeric versions higher than numeric versions

### ISSUE THREAD

**Issue by adamjstewart** [spack/spack#8421]
See #1975, #1979, and #1983 for the long, bloody history of this conflict.

I took a look through all of the non-numeric versions currently in Spack. Here are the most common names:
```
 180 develop
  63 master
   8 xsdk-0.2.0
   3 head
   2 xsdk-0.3.0
   2 git (secretly the master branch)
```
There's a problem here. Aside from the 180 `develop` versions, all of these other versions are being sorted incorrectly. Currently, all non-numeric versions (aside from `develop` which gets special treatment) are considered to be older than all numeric versions. But does this actually make sense? Why would someone add a version control branch to Spack if it wasn't being kept up-to-date? 

I'm sick of telling people to rename the version to `develop` even though it's the `master` branch. What if someone wants to add both `develop` and `master`? Do we tell them to expect concretization bugs with `master` and that there's nothing we can do about it?

In #8262, @bvanessen mentioned that he frequently uses `@local` during development with `spack setup`. Other people use `@system` for external packages. Currently, these versions are being sorted incorrectly.

I have two proposals that could be used to solve this dilemma:

## 1. Sort all non-numeric versions higher than numeric versions

I propose that we sort non-numeric versions higher than numeric versions. We would still install the newest numeric version by default, so there would be no change on the user-side. The only difference is that something like `depends_on('foo@1.2:')` would consider `master` or `head` to be newer than version 1.2.

This proposal doesn't suffer from the problems mentioned in the aforementioned PRs, where there were complaints about having too much special logic.

## 2. Add special cases for `master` and friends

I first proposed this [here](https://github.com/spack/spack/issues/1975#issuecomment-252652273), but we could have other aliases to `develop` that form a hierarchy like so:

| Tier | Version |
| --- | --- |
| Tier 1 | develop/dev/latest |
| Tier 2 | master/head/trunk |
| Tier 3 | numeric versions |
| Tier 4 | non-numeric versions not listed above |

This proposal has the added benefit that Spack would understand that `develop` is newer than `master`. It would also keep us safe from weird unexpected versions that aren't actually new, but from what I've seen these are rare enough to ignore. We could add additional aliases as we encounter them.

_Precedes to reopen can of worms_

@davydden @tgamblin

**Comment by davydden** [comment]
I would suggest avoid using `numeric` / `non-numeric` labelling, because there is really no need to refer to this. IMO it's more confusing. https://github.com/spack/spack/pull/1983 reworked (long time ago) version comparison to avoid this labelling because all we do is split `x.1.z` into `x`, `1`, `z` and compare each with the same split of another version.
In other words, we only need to decide how to compare `strings` and `numbers`. 

Also we avoid confusing about what is actually numeric? Is `1.a` numeric? Is `a.1` numeric? Is `system` numeric? And so on. We really don't need to, see https://github.com/spack/spack/pull/1983#discussion_r132284563


> I'm sick of telling people to rename the version to develop even though it's the master branch.

Merging this almost two-year-old PR https://github.com/spack/spack/pull/1983 would avoid this problem altogether.


> Currently, these versions are being sorted incorrectly.

On this account I disagree. How can spack know what `local` or `system` means? Say if `system` corresponds to `intel-mpi@2017.1`, then why not using `2017.1` in the first place? 
Whereas if `local` means it's something you are **developing**, then `develop.local` would sort this issue and be very descriptive.

**Comment by adamjstewart** [comment]
> I'm sick of telling people to rename the version to develop even though it's the master branch.
> > Merging this almost two-year-old PR #1983 would avoid this problem altogether.

Are you suggesting that we tell people to name the version `develop.master`? Because that's worse than the current situation 😆

**Comment by davydden** [comment]
> Are you suggesting that we tell people to name the version develop.master? Because that's worse than the current situation 😆

No, I don’t! It does address ‘master’, ‘trunk’, etc! Just check unit tests

**Comment by adamjstewart** [comment]
Oh right, I forgot that you added the hierarchical thing, then removed it, then added more `develop` aliases. I would be happy with #1983 as is (although I secretly liked the hierarchical thing better).

**Comment by davydden** [comment]
> although I secretly liked the hierarchical thing better

the code is hierarchical, I just did not have tests for that. Now I added: https://github.com/spack/spack/pull/1983/commits/90decb11dffcf531f1605ac01cf3f496e240cf46

**Comment by adamjstewart** [comment]
Thanks @davydden, sorry I didn't read the code more carefully, it's been a while. You may want to update the PR title/description, that PR has expanded a lot in scope since it was first conceived.

**Comment by citibeth** [comment]
I recommend that the user can give numbers (tuples) to non-numeric versions, which tells Spack how to sort them compared to the numeric versions.  This is basically using a Key Function to sort versions, which is very Pythonic:

https://wiki.python.org/moin/HowTo/Sorting#Key_Functions

**Comment by becker33** [comment]
@citibeth to the best of my knowledge python only uses keys for sorting, and not for comparisons. I was able to find a proposal for a `__key__` function that would automatically generate all of the comparators, but it was not a PEP and has not been implemented. So I think the "Pythonicness" of using keys in this instance is up for debate.

I'm currently reviewing #1983, and I don't think it is fully incompatible with your request for an `effective` keyword in the `version` directive. We could evaluate adding that option independently.

However, I do think the `effective` keyword will not cover all of the cases we want to eventually. For example, release candidates. Although not handled in #1983, we want the behavior `Version('1.2.2.head') < Version('1.2.3-rc1') < Version('1.2.3')`. While this technically can be rendered numerically, any schema for applying numbers to this ordering is complicated, and relies on both an `infinity` token and a `subtract from the previous` token. I think this is more accurately represented by a combination of #1983 and a future PR to add `rc` and `rc-like` tokens to the version comparators than it is by the `effective` keyword.

**Comment by citibeth** [comment]
I think ordering should be defined as reverse of the order they appear in
the recipe.  Putting an ordering on a small and finite set of things should
not be as hard as we’ve made it out to be.

**Comment by becker33** [comment]
I disagree. We have a clear separation of concepts between the Spec and the Package. Which versions are listed in the recipe is part of the Package. The generic ordering of versions is part of the Spec. This may seem pedantic, but it matters for features like `spack version` and `spack checksum`, and especially for the ability to do a best-effort install of versions that do not appear in the recipe.

**Comment by citibeth** [comment]
@becker33 @tgamblin 

There are competing priorities here, and I think we're making the wrong decision on which priorities are more important.  I see the following issues here:

1. Some packages have strange versions that are not easily handled, or sorted correctly, by Spack.

2. Because Spack is automated, checksums are important for security purposes (and for selling Spack to HPC IT departments).

3. Desire to support the following features:
 a) `spack versions`: list available versions of a package
 b) `spack checksum`: checksum available versions of a package
 c) Install versions that do not appear in the recipe.

My analysis of these competing needs is as follows:

* (3c) conflicts directly with (2).  (2) is a more important feature to maintain and enforce in Spack, whereas (3c) is rarely used in production, and by definition cannot  be supported.  The remedy to an unlisted version that doesn't exist is also easy: just add it to the package!

* `spack versions` doesn't intrinsically need to sort versions.  Therefore, it can continue as before: make a best effort at sorting the versions that works in 95% of cases.  Moreover, `spack versions` is NOT an essential part of Spack at all.  Most people already know how to find available versions of a package (i.e. look on the package's website).  That method will always be more robust and trustworthy than `spack versions` --- which depends on having the correct root URL in place, and knowing that all available versions can be downloaded through that URL, and probably a few other assumptions as well.

* `spack checksum` also doesn't need to know how to sort versions.  At the very least, you run it on a single version, and it gives you back a single checksum.  If it does need to sort versions, it can do so on a best effort fashion.  Or in the case of unusual version strings, you could add them to the recipe (with blank checksums) and THEN run `spack checksum`.  In any case, `spack checksum` is also not a core piece of Spack functionality.  I've created plenty of packages without ever using this command.

My conclusion is we are using hard-to-support arguments about ancillary Spack features to justify continued dysfunction on some core Spack functionality (i.e. knowing how to sort versions that HAVE been added to the recipe).  In the meantime, the version-sorting code as it currently stands is unnecessarily complex; certainly not what I would want if I were designing this from scratch, and more Java-like than Pythonic.  Even if you haven't looked at that code, it should be clear that the CORE PROBLEM --- creating an ordering on a finite set of items --- is really simple.

I'm not wedded to the idea that version ordering should be determined by the ordering within the package.  But at least it's a simple solution to a simple problem --- as opposed to an overly complicated solution to a simple problem that doesn't solve all instances of the problem.  Variants on it might be:

1. Keep the current set-up, unless a directive is added to the file indicating that versioning is to be taken based on ordering in the file.

2. Use the Python "key function" idea for sorting, i.e. show Spack how to sort things it doesn't know how to sort (unusual version strings) by replacing them with things it DOES not how to support (say, tuples, i.e. well-behaved dot-separated version strings).  (This idea was discussed in a different Issue).

I would vote for (1).  Although it doesn't simplify the code that's already there, at least it's not disruptive.  And by definition it covers all cases not currently covered.

Along with that, I would want a range operator that does NOT include endpoints.  I.e. instead of saying something like `xyz@3:3.9999`, we can say (for example) `xyz@3::4`, meaning everything up to BUT NOT INCLUDING version 4.  This is important when the only versions you can mention are those that exist.

**Comment by citibeth** [comment]
I'm re-opening because #1983 only solves this problem in some cases; it doesn't even solve all open, reasonable cases people already want to try.  Show me a system that correctly orders all versions of Windows, and then I'll be satisfied the problem is solved.  For reference, here they are in order:

```
Windows 1.0 - 2.0
Windows 3.0 – 3.1
Windows 95
Windows 98
Windows ME - Millennium Edition
Windows NT 31. - 4.0
Windows 2000
Windows XP
Windows Vista
Windows 7
Windows 8
Windows 10
```
(Why is `windows@nt31` < `windows@nt4.0`???)

**Comment by adamjstewart** [comment]
I'm re-closing this because #1983 solves the problem initially reported in my issue. If anyone would like to open another issue to offer further suggestions, feel free.
